esp-hal-common/src/dma/mod.rs

```diff
@@ -691,37 +699,42 @@ where
             let chunk_size = usize::min(CHUNK_SIZE, len - processed);
             let last = processed + chunk_size >= len;
 
-            descriptors[descr + 1] = data as u32 + processed as u32;
-
-            let mut dw0 = &mut descriptors[descr];
+            // buffer flags
+            let dw0 = &mut descriptors[descr];
 
+            // FIXME: why do we need to set 1 for every descriptor in circular mode?
```

Comment:
I think I remember now.

In circular mode we need to know which descriptor has been already processed (obviously). Then we know which part of the buffer can get rewritten.

In theory there might be other ways to do it but back then we tried a few and most of them got it working on some chips but not working on others.

I know the TRM says something like "Specifies whether this descriptor is the last descriptor in the list" and that is just how we use it for non-circular transfers. Given that sentence from the TRM might be confusing when setting it for every descriptor.

However, we can in fact mark every descriptor as EOF and hardware will assert the SUC_EOF_INT for each descriptor marked as EOF (end-of-frame).

IIRC at least on some chips OUTLINK_DSCR_ADDR was only updated when that interrupt got asserted. And that is what we check to know which descriptor is done. I think the TRM doesn't say that but that was the take-away from observation/debugging.

Even if the last paragraph wouldn't be true, SUC_EOF_INT let us know when we need to check OUTLINK_DSCR_ADDR

Reply:
Thanks! I've replaced the FIXME with something that I hope makes sense. I assumed `suc_eof` stops the transfer, it wasn't obvious to me that the hardware instead uses the next descriptor address as its end marker.

Reply:
The TRM could probably do a better job explaining this detail (and I could have made a better job commenting it in the code). Took me some time to discover this, too

Reply:
agreed. anything else I should do here?